tests/main_test.go

```diff
@@ -0,0 +1,258 @@
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/canonical/pebble/internals/servicelog"
+)
+
+// TestMain builds the pebble binary before running the integration tests.
+func TestMain(m *testing.M) {
+	goBuild := exec.Command("go", "build", "-o", "../pebble", "../cmd/pebble")
+	if err := goBuild.Run(); err != nil {
+		fmt.Println("Cannot build pebble binary:", err)
+		os.Exit(1)
+	}
+
+	exitCode := m.Run()
+	os.Exit(exitCode)
+}
+
+func createLayer(t *testing.T, pebbleDir string, layerFileName string, layerYAML string) {
+	t.Helper()
+
+	layersDir := filepath.Join(pebbleDir, "layers")
+	err := os.MkdirAll(layersDir, 0o755)
+	if err != nil {
+		t.Fatalf("Cannot create layers directory: pipe: %v", err)
+	}
+
+	layerPath := filepath.Join(layersDir, layerFileName)
+	err = os.WriteFile(layerPath, []byte(layerYAML), 0o755)
+	if err != nil {
+		t.Fatalf("Cannot create layers file: %v", err)
+	}
+}
+
+func createIdentitiesFile(t *testing.T, pebbleDir string, identitiesFileName string, identitiesYAML string) {
+	t.Helper()
+
+	identitiesPath := filepath.Join(pebbleDir, identitiesFileName)
+	if err := os.WriteFile(identitiesPath, []byte(identitiesYAML), 0o755); err != nil {
+		t.Fatalf("Cannot create layers file: %v", err)
+	}
+}
+
+func pebbleRun(t *testing.T, pebbleDir string, args ...string) <-chan servicelog.Entry {
+	t.Helper()
+
+	logsCh := make(chan servicelog.Entry)
+
+	cmd := exec.Command("../pebble", append([]string{"run"}, args...)...)
+	cmd.Env = append(os.Environ(), "PEBBLE="+pebbleDir)
+
+	t.Cleanup(func() {
+		err := cmd.Process.Signal(os.Interrupt)
+		if err != nil {
+			t.Errorf("Error sending SIGINT/Ctrl+C to pebble: %v", err)
+		}
+		cmd.Wait()
+	})
+
+	stdoutPipe, err := cmd.StdoutPipe()
+	if err != nil {
+		t.Fatalf("Cannot create stdout pipe: %v", err)
+	}
+	stderrPipe, err := cmd.StderrPipe()
+	if err != nil {
+		t.Fatalf("Cannot create stderr pipe: %v", err)
+	}
+
+	err = cmd.Start()
+	if err != nil {
+		t.Fatalf("Error starting 'pebble run': %v", err)
+	}
+
+	done := make(chan struct{})
+	go func() {
+		defer close(logsCh)
+
+		readLogs := func(parser *servicelog.Parser) {
+			for parser.Next() {
+				if err := parser.Err(); err != nil {
+					t.Errorf("Cannot parse Pebble logs: %v", err)
+				}
+				select {
+				case logsCh <- parser.Entry():
+				case <-done:
+					return
+				}
+			}
+		}
+
+		// Both stderr and stdout are needed, because pebble logs to stderr
+		// while with "--verbose", services otuput to stdout.
+		stderrParser := servicelog.NewParser(stderrPipe, 4*1024)
+		stdoutParser := servicelog.NewParser(stdoutPipe, 4*1024)
+
+		// Channel to signal completion and close logsCh
+		done := make(chan struct{})
+		defer close(done)
+
+		go readLogs(stderrParser)
+		go readLogs(stdoutParser)
+
+		// Wait for both parsers to finish
+		<-done
+		<-done
+	}()
+
+	return logsCh
+}
+
+func waitForLogs(logsCh <-chan servicelog.Entry, expectedLogs []string, timeout time.Duration) error {
```

Comment:
After applying my comments in the tests, we may not need this to wait for more than one log at a time, so perhaps we can simplify and just have it wait for a single log? Then we can also check the service name is "pebble" or "svc1" or whatever it should be. I also think that like the other test helpers, this should take `t` and fail, rather than return an error.

```go
func waitForLog(t, logsCh <-chan servicelog.Entry, serviceName, expectedLog string, timeout time.Duration) { ... }

// Example usage:
waitForLog(t, stderrLogs, "pebble", "Started daemon", 3*time.Second)
```

Reply:
After applying all comments above, the only use case of `waitForLog` would be to wait for the stdout that comes from `--verbose`.

Another use case would be to wait for "Started daemon", to make sure the daemon is already started before checking other things (like running `pebble identity` commands), but this could be solved by waiting for 100 milliseconds, just as in `TestHold`.

So, shall we change it to something like `waitForStdout`, and make `pebbleRun` only return one channel for stdout, not stderr (we are not checking pebble logs anyway if we change waiting for `Started daemon" to sleep)?